### PR TITLE
crypto: Switch to base58 for TransactionDigest

### PR DESCRIPTION
**Pull request by joyqvq**

### CODE REVIEW

**Comment by 666lcz** [comment]
Can we also update https://github.com/MystenLabs/sui/blob/21c26ce6a5d4e3448abd74323e3164286d3deba6/sdk/typescript/src/types/common.ts#L23 ?

**Comment by joyqvq** [comment]
> Can we also update
> 
> https://github.com/MystenLabs/sui/blob/21c26ce6a5d4e3448abd74323e3164286d3deba6/sdk/typescript/src/types/common.ts#L23
> 
> ?

added - there is no e2e search test for tx digest thats why nothing fails on ci. just added a e2e test for that

**Review by kchalkias — COMMENTED** [review]
Changes seem fine, are we sure nothing else breaks? wallet, capy, and others that need to be informed? @damirka ?

**Comment by damirka** [comment]
Thanks for pinging me @kchalkias! Is the change only for tx digest (and not object digest)? If so, then we’re good. If it affects object digest then we’ll need changes in the bcs implementation as well.

**Comment by joyqvq** [comment]
> Thanks for pinging me @kchalkias! Is the change only for tx digest (and not object digest)? If so, then we’re good. If it affects object digest then we’ll need changes in the bcs implementation as well.

correct not object digest. although it should not impact bcs regardless, since bcs is concerning struct -> bytes not bytes -> string, base58 would only affect the latter.

File: sdk/typescript/src/cryptography/hash.ts [hunk 1]
```diff
@@ -20,5 +20,5 @@ export function sha256Hash(typeTag: string, data: Uint8Array): string {
 
   hash.update(dataWithTag);
 
-  return Buffer.from(hash.hex(), 'hex').toString('base64');
+  return Buffer.from(hash.hex(), 'hex');
```
Comment (Jordan-Mysten) [inline sdk/typescript/src/cryptography/hash.ts:23]: When did we re-add buffer to the SDK? We should never use the native Buffer type.

File: sdk/typescript/src/types/common.ts [hunk 2]
```diff
@@ -13,8 +13,9 @@ import {
 import { sha256Hash } from '../cryptography/hash';
 import { Ed25519PublicKey } from '../cryptography/ed25519-publickey';
 import { Secp256k1PublicKey } from '../cryptography/secp256k1-publickey';
+import { Base64DataBuffer } from '../serialization/base64';
 
-/** Base64 string representing the object digest */
+/** Base58 string representing the transaction digest */
```
Comment (Jordan-Mysten) [inline sdk/typescript/src/types/common.ts:18]: We should support both at the same time while this rolls out.

Comment (joyqvq) [reply sdk/typescript/src/types/common.ts]: good call. in the case of explorer gets deployed before sui, it should parse base64.

File: sdk/typescript/src/cryptography/hash.ts (hunk 1 above)
Comment (joyqvq) [reply sdk/typescript/src/cryptography/hash.ts]: good catch, this was recently introduced, removed.

File: apps/explorer/src/utils/static/mock_data.json [hunk 3]
```diff
@@ -595,7 +595,7 @@
             "readonly": "true",
             "objType": "0x2::coin::Coin<0x2::USD::USD>",
             "data": {
-                "tx_digest": "Da4vHc9IwbvOYblE8LnrVsqXwryt2Kmms+xnJ7Zx5E4=",
```
Comment (joyqvq) [inline apps/explorer/src/utils/static/mock_data.json:598]: mock_data will be deprecated in favor of localnet e2e test, still added here for consistency

File: sdk/typescript/src/cryptography/hash.ts [hunk 4]
```diff
@@ -20,5 +21,5 @@ export function sha256Hash(typeTag: string, data: Uint8Array): string {
 
   hash.update(dataWithTag);
 
-  return Buffer.from(hash.hex(), 'hex').toString('base64');
```
Comment (joyqvq) [inline sdk/typescript/src/cryptography/hash.ts:23]: buffer removed.

File: sdk/typescript/src/types/common.ts [hunk 5]
```diff
@@ -26,17 +26,23 @@ export type ObjectOwner =
...
-  return (
-    new Base64DataBuffer(value).getLength() === TX_DIGEST_LENGTH &&
-    VALID_BASE64_REGEX.test(value)
-  );
+  let buffer;
+  try {
+    buffer = new Base58DataBuffer(value);
+  } catch (e) {
+    try {
+      // For backwards compatibility, we also accept Base64 strings
```
Comment (666lcz) [inline sdk/typescript/src/types/common.ts:39]: I would pass in a `serializationFmt: 'base64' | 'base58'` to this method. The application will choose which one to use depending on API version

File: sdk/typescript/src/types/common.ts [hunk 6]
```diff
@@ -134,7 +140,8 @@ export function generateTransactionDigest(
   const senderSignedDataBytes = bcs
     .ser('SenderSignedData', senderSignedData)
     .toBytes();
-  return sha256Hash('SenderSignedData', senderSignedDataBytes);
+
+  return new Base58DataBuffer(sha256Hash('SenderSignedData', senderSignedDataBytes)).toString();
```
Comment (666lcz) [inline sdk/typescript/src/types/common.ts:144]: For backward compatibility, I think you can pass in a `serializationFmt: 'base58' | 'base64' ` to the function signature of `generateTransactionDigest`. And in line 144, you can do `serializationFmt === 'base58' ? ... : ...)

File: sdk/typescript/src/types/common.ts (hunk 5 above)
Comment (joyqvq) [reply sdk/typescript/src/types/common.ts]: revised

File: sdk/typescript/src/types/common.ts (hunk 6 above)
Comment (joyqvq) [reply sdk/typescript/src/types/common.ts]: revised

**Comment by 666lcz** [comment]
Looks like there's some build error to fix - otherwise looks great to me! Thanks for adding the tests!

**Review by 666lcz — APPROVED** [review]

**Review by kchalkias — APPROVED** [review]
Agree for backwards b64 compatibility. I guess our explorer and wallets (and partners gradually) will prefer the new b58 one and the teams need to know.

**Pull request merged**